hadoop-hdds/rocks-native/src/main/native/SSTDumpTool.cpp

```diff
@@ -0,0 +1,49 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "org_apache_hadoop_hdds_utils_db_managed_ManagedSSTDumpTool.h"
+#include "rocksdb/options.h"
+#include "rocksdb/sst_dump_tool.h"
+#include <string>
+#include "cplusplus_to_java_convert.h"
+#include "Pipe.h"
+#include <iostream>
+
+jint Java_org_apache_hadoop_hdds_utils_db_managed_ManagedSSTDumpTool_runInternal(JNIEnv *env, jobject obj,
+ jobjectArray argsArray, jlong optionsHandle, jlong pipeHandle) {
+    ROCKSDB_NAMESPACE::SSTDumpTool dumpTool;
+    ROCKSDB_NAMESPACE::Options *options = reinterpret_cast<ROCKSDB_NAMESPACE::Options *>(optionsHandle);
+    Pipe *pipe = reinterpret_cast<Pipe *>(pipeHandle);
+    int length = env->GetArrayLength(argsArray);
+    char *args[length + 1];
+    args[0] = strdup("./sst_dump");
+    for (int i = 0; i < length; i++) {
+        jstring str_val = (jstring)env->GetObjectArrayElement(argsArray, (jsize)i);
+        char *utf_str = (char *)env->GetStringUTFChars(str_val, JNI_FALSE);
+        args[i + 1] = strdup(utf_str);
+        env->ReleaseStringUTFChars(str_val, utf_str);
+    }
+    FILE *wr = fdopen(pipe->getWriteFd(), "w");
+    int ret = dumpTool.Run(length + 1, args, *options, wr);
+    for (int i = 0; i < length + 1; i++) {
+        free(args[i]);
+    }
+    fclose(wr);
+    pipe->close();
```

Comment:
Need to close `options` as well. e.g.

https://github.com/facebook/rocksdb/blob/dec144f172165e9ce007aa12fbeaacf748cea5c1/java/src/test/java/org/rocksdb/SstFileReaderTest.java#L112

Reply:
Since `pipe` is also passed in from the Java side. Should Java code be taking care of `pipe` as well? (i.e. don't close `pipe` in C++ code here)

Reply:
we shouldn't close options here as we are passing the already created options to sst dump tool constructor. I did this deliberately.

Reply:
Could `PipeInputStream` Java object could be garbage collected by JVM before C++ code here finished using it? Looks like it could be?

In other words, is there any guarantee in KVM that objects passed through JNI won't be GC'ed?

Reply:
If GC is triggered on PipeInputStream then we close the pipe, it basically means that we are not using the inputstream right & thus wouldn't be needing the pipe.

Reply:
The finalize method has been overriden for this.

Reply:
no we shouldn;t close options here since option is passed to sst dump tool.